util/signature/verifier.go

```diff
@@ -5,64 +5,117 @@ package signature
+	f.Bool(prefix+".accept-missing", DefultFeedVerifierConfig.Dangerous.AcceptMissing, "accept empty as valid signature")
+}
+
+var DefultFeedVerifierConfig = VerifierConfig{
+	AllowedAddresses:   []string{},
+	AcceptBatchPosters: true,
+	Dangerous: DangerousVerifierConfig{
+		AcceptMissing: true,
+	},
+}
+
+var TestingFeedVerifierConfig = VerifierConfig{
+	AllowedAddresses:   []string{},
+	AcceptBatchPosters: false,
+	Dangerous: DangerousVerifierConfig{
+		AcceptMissing: false,
+	},
 }
 
-func NewVerifier(requireSignature bool, authorizedAddresses []common.Address, bpValidator contracts.BatchPosterVerifierInterface) *Verifier {
-	authorizedMap := make(map[common.Address]struct{}, len(authorizedAddresses))
-	for _, addr := range authorizedAddresses {
+func NewVerifier(config *VerifierConfig, bpValidator contracts.BatchPosterVerifierInterface) (*Verifier, error) {
+	authorizedMap := make(map[common.Address]struct{}, len(config.AllowedAddresses))
+	for _, addrString := range config.AllowedAddresses {
+		addr := common.HexToAddress(addrString)
 		authorizedMap[addr] = struct{}{}
 	}
-	return &Verifier{
-		requireSignature: requireSignature,
-		authorizedMap:    authorizedMap,
-		bpValidator:      bpValidator,
+	if bpValidator == nil && config.AcceptBatchPosters {
+		return nil, errors.New("cannot read batch poster addresses")
 	}
+	return &Verifier{
+		config:        config,
+		authorizedMap: authorizedMap,
+		bpValidator:   bpValidator,
+	}, nil
 }
 
 func (v *Verifier) VerifyHash(ctx context.Context, signature []byte, hash common.Hash) (bool, error) {
-	return v.verifyClosure(ctx, signature, func() common.Hash { return hash })
+	return v.verifyClosure(ctx, signature, hash)
 }
 
 func (v *Verifier) VerifyData(ctx context.Context, signature []byte, data ...[]byte) (bool, error) {
-	return v.verifyClosure(ctx, signature, func() common.Hash { return crypto.Keccak256Hash(data...) })
+	return v.verifyClosure(ctx, signature, crypto.Keccak256Hash(data...))
 }
 
-var ErrMissingFeedSignature = errors.New("missing required feed signature")
+var ErrMissingSignature = errors.New("missing required signature")
 
-func (v *Verifier) verifyClosure(ctx context.Context, signature []byte, getHash func() common.Hash) (bool, error) {
-	if len(signature) == 0 {
-		if !v.requireSignature {
+func (v *Verifier) verifyClosureLocal(sig []byte, hash common.Hash) (bool, common.Address, error) {
+	if len(sig) == 0 {
+		if v.config.Dangerous.AcceptMissing {
 			// Signature missing and not required
-			return true, nil
+			return true, common.Address{}, nil
 		}
-
-		return false, ErrMissingFeedSignature
+		return false, common.Address{}, ErrMissingSignature
 	}
 
-	var hash = getHash()
-
-	sigPublicKey, err := crypto.SigToPub(hash.Bytes(), signature)
+	sigPublicKey, err := crypto.SigToPub(hash.Bytes(), sig)
 	if err != nil {
-		return false, errors.Wrap(err, "unable to recover sequencer feed signing key")
+		// nolint:nilerr
+		return false, common.Address{}, nil
```

Comment:
My thinking is instead of a valid bool, we should just return `(common.Address, error)` and return an error here or anytime else the signature is invalid. We can propagate this all the way up to the VerifyData / VerifyHash function. It should make it safer to use as linters will warn you if you ignore an error, but not if you ignore a bool.